web/webserver.go

```diff
@@ -1,182 +1,74 @@
-	// markbates/goth/gothic.GetProviderName can find it.
-	withProvider := func(next echo.HandlerFunc) echo.HandlerFunc {
-		return func(c echo.Context) error {
-			qv := c.Request().URL.Query()
-			qv.Set("provider", c.Param("provider"))
-			c.Request().URL.RawQuery = qv.Encode()
-			return next(c)
+// MuxHandler routes HTTP and gRPC requests.
+func (m *GrpcMultiplexer) MuxHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if m.MuxServer.IsGrpcWebRequest(r) {
+			log.Printf("MUX: got GRPC request: %v", r)
+			m.MuxServer.ServeHTTP(w, r)
+			return
 		}
-	}
-
-	oauth2 := e.Group("/auth/:provider", withProvider, auth.PreAuth(ags.logger, ags.db))
-	oauth2.GET("", auth.OAuth2Login(ags.logger, ags.db))
-	oauth2.GET("/callback", auth.OAuth2Callback(ags.logger, ags.db, ags.scms))
-	e.GET("/logout", auth.OAuth2Logout(ags.logger))
+		next.ServeHTTP(w, r)
+	})
 }
 
-func registerFrontend(e *echo.Echo, entryPoint, public string) {
-	index := func(c echo.Context) error {
-		return c.File(entryPoint)
-	}
-
-	e.GET("/", index)
-	e.GET("/*", index)
-	// TODO: Whitelisted files only.
-	e.Static("/static", filepath.Join(public, "dist"))
-	e.Static("/assets", filepath.Join(public, "assets"))
-}
+// RegisterRouter registers http endpoints for authentication API and scm provider webhooks.
+func (s *QuickFeedService) RegisterRouter(authConfig *auth.AuthConfig, scms *auth.Scms, mux GrpcMultiplexer, public string) *http.ServeMux {
+	// Serve static files.
+	router := http.NewServeMux()
+	assets := http.FileServer(http.Dir(public + "/assets"))
+	dist := http.FileServer(http.Dir(public + "/dist"))
 
-func runWebServer(l *zap.SugaredLogger, e *echo.Echo, httpAddr string) {
-	e.Server.WriteTimeout = writeTimeout
-	e.Server.ReadTimeout = readTimeout
-	e.Server.IdleTimeout = idleTimeout
+	router.Handle("/", mux.MuxHandler(http.StripPrefix("/", assets)))
+	router.Handle("/assets/", mux.MuxHandler(http.StripPrefix("/assets/", assets)))
+	router.Handle("/static/", mux.MuxHandler(http.StripPrefix("/static/", dist)))
 
-	e.TLSServer.ReadTimeout = readTimeout
-	e.TLSServer.WriteTimeout = writeTimeout
-	e.TLSServer.IdleTimeout = idleTimeout
+	// Register auth endpoints.
+	callbackSecret := rand.String()
+	router.HandleFunc("/auth/", auth.OAuth2Login(s.logger, authConfig, callbackSecret))
+	router.HandleFunc("/auth/callback/", auth.OAuth2Callback(s.logger, s.db, authConfig, scms, callbackSecret))
+	router.HandleFunc("/logout", auth.OAuth2Logout(s.logger))
 
-	srvErr := e.Start(httpAddr)
-	if srvErr == http.ErrServerClosed {
-		l.Warn("shutting down the server")
-		return
-	}
-	l.Fatal("failed to start server", zap.Error(srvErr))
-	// TODO(meling) pretty sure the following lines are unreachable; check
-
-	quit := make(chan os.Signal, 1)
-	signal.Notify(quit, os.Interrupt)
-	<-quit
+	// Register hooks.
+	ghHook := hooks.NewGitHubWebHook(s.logger, s.db, s.runner, s.bh.Secret)
```

Comment:
Here we use `s.bh.Secret` which is read from `WEBHOOK_SECRET` env var. However, on Line 64 `callbackSecret` is generated. Should these be the same? Recall also #480.

We use this struct below; perhaps it would make sense to add the `oauth2.Config` to it... and use that to pass as input to the `OAuth2` funcs. That said, if we do that, then is should be given a better name.

```go
// BaseHookOptions store the application's callback url and secret
type BaseHookOptions struct {
	BaseURL string
	Secret  string
	Config *oauth2.Config
}
```

Reply:
The difference here is that `callbackSecret` is used in the user authentication (in `OAuth2Login` we pass the secret to GitHub, and in `OAuth2Callback` we check the secret in the response from GitHub to make sure the response has not been forged. So it is better to randomly generate `callbackSecret` instead of reading it from env for security reasons.

Webhook secret, on the other hand, is constant and set when the webhook is created (currently the same secret for each new course). This means the `s.bh.secret` will never change, so it makes sense to read it from the environment.